### PR TITLE
Feature/give weight values to predicates

### PR DESCRIPTION
**Pull request by richard0000**
# What was done?

### Add weight to predicates
- Each predicate now has a "weight" property for assigning a weight to each concept for calculating the alignment score (%) in the display of the mapping.

### Improve parser to be more explicit
- When a user is trying to map twice to the same domain using the same file, we recognize it by looking at the URI's and throw an error to warn the user.

###     Show percentage progress bar on spine property card
- Perform all the calculations for the alignment weights to show a progress bar with a percentage on the spine property card.

# Screenshots
image

### CODE REVIEW

File: app/controllers/api/v1/mapping_terms_controller.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def authorize_with_policy
   # @return [ActiveRecord::Relation]
   ###
   def filter
-    terms = MappingTerm.all
+    terms = MappingTerm.joins(:mapping).where(mappings: {status: :mapped})
 
     terms = MappingTerm.where(spine_term_id: params[:spine_term_id]) if params[:spine_term_id].present?
```
Comment (excelsior) [inline app/controllers/api/v1/mapping_terms_controller.rb:49]: Shouldn't the `status: :mapped` condition be applied here as well?

File: app/javascript/components/mapping/MappingPreview.jsx [hunk 2]
```diff
@@ -19,6 +19,7 @@ import { setVocabularies, unsetVocabularies } from "../../actions/vocabularies";
 import createVocabulary from "../../services/createVocabulary";
 import { vocabName } from "../../helpers/Vocabularies";
 import UploadVocabulary from "../mapping-to-domains/UploadVocabulary";
+import Pluralize from "pluralize";
```
Comment (excelsior) [inline app/javascript/components/mapping/MappingPreview.jsx:22]: Let's keep function names starting with lowercase letters for the sake of consistency.

File: app/services/processors/mappings.rb [hunk 3]
```diff
@@ -34,8 +36,8 @@ def self.create(specification, user)
     # @return [Array]
     ###
     def self.create_mapping_terms mapping
-      spine = mapping.spine
-      spine.terms.each do |term|
+      terms = 0
+      mapping.spine.terms.each do |term|
         # Do not create this term if there's already one with the same uri
         next if MappingTerm.find_by(uri: term.desm_uri)
```
Comment (excelsior) [inline app/services/processors/mappings.rb:42]: Better use `exists?` instead of `find_by`, since there's no need to instantiate a term.

**Review by excelsior — COMMENTED** [review]
Thanks! Could you please address the comments?

File: app/javascript/components/mapping/MappingPreview.jsx (hunk 2 above)
Comment (richard0000) [reply app/javascript/components/mapping/MappingPreview.jsx:22]: It's a dependency, the name comes with it.

**Pull request merged**